Add blog post: GitOps Best Practices I Wish I Had Known Before (#17468)
pulumi/docs: content/blog/gitops-best-practices-i-wish-i-had-known-before/index.md

meta_image: meta.png

authors:
- engin-diri

tags:
- gitops
- kubernetes
- best-practices
- argocd
- devops

social:
  twitter: |
    GitOps Best Practices I Wish I Had Known Before — hard-won lessons from production, the IaC-to-GitOps bridge, and why pragmatism beats dogma every time.
    <link>
  linkedin: |
    GitOps Best Practices I Wish I Had Known Before

    Getting started with GitOps? It can feel like you're trying to herd cats through a YAML factory. After years of running GitOps in production across dozens of clusters, I've compiled the essential best practices that could save you from the mistakes I made.

    From repo structure to secret management, drift detection to the IaC-to-GitOps bridge — this is the curated cheat sheet I wish I'd had from Day 1.
    <link>
---

Getting started with GitOps can feel like trying to herd cats through a YAML factory while the factory is on fire. After years of running GitOps workflows in production across dozens of clusters, I've collected a list of best practices that could save you from the mistakes I made. Think of this as the GitOps cheat sheet I wish I'd had from Day 1.

<!--more-->

> If you're not familiar with the formal definition, the [OpenGitOps](https://opengitops.dev/) project distills it into four principles: declarative desired state, versioned and immutable storage, automatic pulling, and continuous reconciliation. Those principles define the "what." This post is about the "how" — the practical lessons that make or break a GitOps implementation.

Reviewer: This feels like it wants to be a list:

```suggestion
If you're not familiar with the formal definition, the [OpenGitOps](https://opengitops.dev/) project distills it into four principles: 

* Declarative desired state
* Versioned and immutable storage
* Automatic pulling
* Continuous reconciliation

But those principles only define the _what_. This post is about the _how_ — the practical lessons that make or break a GitOps implementation.
```

{{% notes type="info" %}}
Going to [KubeCon Europe 2026](https://www.pulumi.com/kubecon/)? Tame Kubernetes complexity with code.

Pulumi demos, platform engineering talks, AI-powered Neo in action, and yes, free plushies. [Booth 784](https://www.pulumi.com/kubecon/).
{{% /notes %}}

> 

Author: Done. Removed the KubeCon callout entirely.

> In this post, I'll walk you through the GitOps best practices I've picked up from production experience, community talks, and more than a few late-night incident calls. Whether you're just getting started with GitOps or looking to level up an existing setup, these tips should help you avoid the potholes.

Reviewer: ```suggestion
In this post, I'll walk you through the GitOps best practices I've picked up from production experience, community talks, and more than a few late-night incident calls. Whether you're just getting started with GitOps or looking to level up, these tips should help you avoid the potholes.
```

> **Key takeaways:** Git is your only source of truth — no manual edits. Use pull-based, declarative deployments with directory-per-environment structure. Validate every change in CI before it hits your cluster. Manage secrets outside of Git. Bridge your IaC and GitOps layers instead of choosing one. And above all, be pragmatic — even the biggest GitOps adopters deviate when the situation calls for it.

Author: Done. Removed the key takeaways paragraph entirely.

1. [Progressive delivery: don't YOLO to production](/blog/gitops-best-practices-i-wish-i-had-known-before/#11-progressive-delivery-dont-yolo-to-production)
1. [Policy-as-code: your automated guardrails](/blog/gitops-best-practices-i-wish-i-had-known-before/#12-policy-as-code-your-automated-guardrails)
1. [Secrets don't belong in Git (ever)](/blog/gitops-best-practices-i-wish-i-had-known-before/#13-secrets-dont-belong-in-git-ever)
1. [Bridge your IaC and GitOps (don't choose one)](/blog/gitops-best-practices-i-wish-i-had-known-before/#14-bridge-your-iac-and-gitops-dont-choose-one)
1. [Be pragmatic, not dogmatic](/blog/gitops-best-practices-i-wish-i-had-known-before/#15-be-pragmatic-not-dogmatic)
1. [Invest in observability (GitOps doesn't replace monitoring)](/blog/gitops-best-practices-i-wish-i-had-known-before/#16-invest-in-observability-gitops-doesnt-replace-monitoring)

{{< figure alt="Drake meme: rejecting kubectl apply, approving git push" src="gitops-meme.png" caption="The GitOps mindset in a nutshell." width=100% >}}

## 1. Git is your single source of truth (no, really)

> This is the foundational principle of GitOps, and yet it's the one people break first. Every piece of your environment's desired state lives in a Git repository. No exceptions. No "I'll just fix it real quick with `kubectl edit`." No "let me patch this configmap by hand because the PR process takes too long."

Reviewer: > This is the foundational principle of GitOps, and yet it's the one people break first

Which of the four principles is this one? And when they break it, how do they brea

```suggestion
This is the foundational principle of GitOps: Every piece of your environment's desired state lives in a Git repository. No exceptions. No "I'll just fix it real quick with `kubectl edit`." No "let me patch this configmap by hand because the PR process takes too long."
```

{{% notes type="info" %}}
Pro Tip: Enable branch protection rules on your GitOps repos from Day 1. This prevents anyone from pushing directly to main and bypassing the review process. Future you will thank present you during the next audit.
{{% /notes %}}

## 2. Declarative over imperative, always

If you're still running sequences of `kubectl create`, `kubectl patch`, and `kubectl delete` commands to manage your cluster, you're doing GitOps on hard mode. Declarative means you define *what* you want the end state to look like, not the step-by-step instructions to get there.

Think of it like ordering at a restaurant. Imperative: "Go to the kitchen, grab flour, knead dough, preheat the oven to 200 degrees, shape the pizza, add sauce..." Declarative: "One margherita pizza, please." Let the system figure out how to make it happen.

- Your manifests describe the end result. The GitOps operator reconciles reality to match.
- It's idempotent by design. Apply the same manifest ten times, get the same result. No side effects, no surprises.

> - Rollbacks are easier: just revert a commit. The operator sees the previous desired state and reconciles. A caveat though: this works cleanly for stateless, declarative resources. Database schema migrations, CRD version changes, persistent volume modifications, and rotated secrets don't revert by just reverting a manifest, so plan those rollback paths separately.

Reviewer: ```suggestion
- Rollbacks are easier: just revert a commit. The operator sees the previous desired state and reconciles. A caveat though: this only works for _stateless_ resources; database schema migrations, CRD version changes, persistent volume modifications, and rotated secrets don't always revert cleanly by rolling back to a previous commit. Be sure to plan rollbacks involving stateful resources carefully.
```

Author: Done. Rewrote to clarify this only works for stateless resources and that rollbacks involving stateful resources need careful planning.

{{% notes type="info" %}}
Pro Tip: Set your reconciliation interval to something aggressive like 1 minute. A team running FluxCD in production for a German automotive manufacturer found that a 1-minute interval meant they always knew exactly which version was deployed. A word of caution at scale though: with 200+ applications polling every minute, you can blow through GitHub's API rate limits (5,000 requests/hour for authenticated users) and put real pressure on the Kubernetes API server. Consider tiering intervals by environment, maybe 1 minute for production and 5 minutes for dev/staging.

Or better yet, set up [webhook receivers](https://argo-cd.readthedocs.io/en/latest/operator-manual/webhook/). Both ArgoCD and [FluxCD](https://fluxcd.io/flux/guides/webhook-receivers/) support incoming webhooks from GitHub, GitLab, and Bitbucket. Your Git provider pings the GitOps controller on every push, so reconciliation kicks off in seconds instead of waiting for the next poll. That alone kills most of the API rate limit pain at scale. You can then relax polling to a 5 or 10 minute fallback for the rare case where a webhook doesn't fire.
{{% /notes %}}

> {{< related-posts >}}

Reviewer: This feels very out of place here; we want them to read the post, not get distracted mid-read by other posts. It belongs at the end.  

I'll find a standard place for it in the template.

- When config tweaks start firing your full CI pipeline unnecessarily, it's time to split.
- When different teams need different access levels (not everyone who pushes app code should modify production deployment config), it's time to split.
- When you're running microservices with independent release cadences that step on each other, it's time to split.

If none of those apply yet, don't split. You'll know when the pain arrives.

{{% notes type="info" %}}

> Pro Tip: When you do split, a common pattern is two repos: one for application source code, one for deployment configuration (Helm charts, Kustomize overlays, environment-specific values). Some larger orgs go further with a third repo for environment overrides, following the "fleet repo, infra repo, apps repo" model from the FluxCD maintainers. Either way, pair it with image update automation (Flux Image Automation Controller or ArgoCD Image Updater) so image tag bumps across repos don't turn into manual toil.

Reviewer: >the "fleet repo, infra repo, apps repo" model from the FluxCD maintainers

This feels like it could use a link. Got one?

>Either way, pair it with image update automation (Flux Image Automation Controller or ArgoCD Image Updater) so image tag bumps across repos don't turn into manual toil.

Is this something the person reading an article like this would know how to do? if not (or even if so), maybe another link?

{{% /notes %}}

## 5. Use directories, not branches, for environments

This is the number one anti-pattern I encounter in the wild. Teams create a `dev` branch, a `staging` branch, and a `prod` branch, then cherry-pick or merge between them for promotions. It sounds logical, but it's a trap. And as Conway's Law reminds us, your repo structure will end up mirroring your org structure whether you plan for it or not, so it's worth being intentional about this from the start.

Branches diverge. Cherry-picks get missed. ConfigMaps and secrets meant for dev sneak into staging. Before you know it, your environments have drifted apart and you can't tell what's actually different between them.

> - A folder structure like `environments/dev/`, `environments/staging/`, `environments/prod/` makes differences visible in a single `diff` command.

Reviewer: Question that seems relevant here (as a Kubernetes non-expert): Where and how does this practice overlap with traditional Pulumi stack-based config?

- Moving a change from dev to prod is just copying or updating files across directories, reviewed via a standard pull request.
- You'll never accidentally skip a commit or introduce an environment-specific change where it doesn't belong. No more cherry-pick roulette.

> One nuance worth calling out: the [rendered manifests pattern](https://akuity.io/blog/the-rendered-manifests-pattern) does use per-environment branches, and you'll see more teams adopting it. But it's not the same anti-pattern. With rendered manifests, your main branch stays the only place humans write config (using directories, just like above). CI renders the final YAML and pushes it to read-only environment branches like `env/staging` or `env/prod`. ArgoCD points at those output branches instead of the source. The branches are machine-generated artifacts — nobody commits to them directly and they never merge back to main. That's a completely different animal from "let's use the `staging` branch for staging config and cherry-pick to prod." If you want to try it, ArgoCD's built-in [source hydrator](https://argo-cd.readthedocs.io/en/latest/user-guide/source-hydrator/) and [Kargo Render](https://github.com/akuity/kargo-render) can automate the rendering for you.

Reviewer: > does use per-environment branches, and you'll see more teams adopting it

I don't understand. So are the per-environment branches mentioned here managed by by machines (e.g., Argo), and not humans? If that's the case, is this worth mentioning, since this is about human-managed branches (and in this case, the humans are all using `main`)?

│   └── kustomization.yaml
├── environments/
│   ├── dev/
│   │   ├── kustomization.yaml   # replicas: 1, limits: 256Mi
│   │   └── patches/
│   ├── staging/
│   │   ├── kustomization.yaml   # replicas: 2, limits: 512Mi
│   │   └── patches/
│   └── prod/
│       ├── kustomization.yaml   # replicas: 3, limits: 1Gi
│       └── patches/
```

Each environment's `kustomization.yaml` references the shared base and applies only its overrides. Differences between environments are always one `diff` away.

{{% notes type="info" %}}
Pro Tip: Use Kustomize overlays to keep a shared base and per-environment patches. Your base directory holds the common manifests, and each environment directory contains only the differences. This minimizes duplication while keeping environment boundaries obvious. And in practice, it's not "Kustomize or Helm" — most teams use both in tandem. Kustomize patches work great when you know the values ahead of time (replicas, resource limits). Helm parameterization shines when values depend on the destination cluster and aren't known until deploy time, like an Ingress FQDN or a cloud-specific endpoint. Use Kustomize to keep things DRY, Helm to keep things flexible.

> {{% /notes %}}

Author: Done. Moved the code block inline with the Kustomize explanation, trimmed the pro tip, added a Kustomize link, and replaced the old Kustomize/Helm pro tip with the Pulumi stack config mention.

## 6. Trunk-based development or bust

Long-lived feature branches in a GitOps repo are a recipe for merge conflicts and stale configuration. Your GitOps repository should follow trunk-based development: one main branch, short-lived feature branches, and frequent merges.

> The longer a branch lives, the further it drifts from reality. And in a GitOps context, "drifting from reality" means your environments are out of sync with what people think they look like. That's how incidents happen.

Reviewer: Merge conflicts yes, but "stale configuration"? 🤔 Maybe something more general:

```suggestion
Long-lived feature branches in a GitOps repo are a recipe for sadness; The longer a branch lives, the further it drifts from reality. Your GitOps repository should follow trunk-based development: one main branch, short-lived feature branches, and frequent merges.
```

- Create a branch, make the change, open a PR, get it reviewed, merge. The whole cycle should take hours, not days or weeks.
- Avoid the `develop` / `release` / `main` branching model for GitOps repos. It adds complexity without value here.
- The more often you merge to main, the smaller each change is, and the easier it is to review and roll back.

If you're using the rendered manifests pattern from the previous section, this still applies to your source branch. The environment branches (`env/dev`, `env/prod`) are CI-generated output, not places where humans do work. Your trunk-based workflow stays on main. The environment branches just happen to exist alongside it, and that's fine because nobody is branching off them or merging them back.

{{% notes type="info" %}}
Pro Tip: If your GitOps repo has branches older than a week that aren't machine-generated environment branches, treat it as a code smell. Set up automated reminders or CI checks that flag stale branches.

> {{% /notes %}}

Author: Done. Section removed entirely as part of the cut to 12 tips.

## 7. Every change gets a PR review (no exceptions)

> A sobering stat from a production talk about running FluxCD at a German automotive manufacturer: the number one cause of failed deployments was missing PR reviews. Not bad YAML. Not misconfigured Helm charts. Missing reviews.

Reviewer: Slightly odd not to include a link to this talk, given the detailed reference ("a production talk about running FluxCD at a German automotive manufacturer"). Is that really relevant? If we have the talk, let's link to it, but if not, we should remove that extra bit, as it doesn't really help to strengthen the point (which is otherwise pretty clear).

 
```suggestion
A sobering stat I heard recently: the number one cause of failed deployments was missing PR reviews. Not bad YAML, not misconfigured Helm charts. Missing reviews.
```

> When someone pushes directly to main without review, there's no second pair of eyes to catch the typo in the resource limit, the missing namespace, or the wrong image tag. PR reviews are your last check before a change hits your cluster.

Reviewer: ```suggestion
When someone pushes directly to main without review, there's no second pair of eyes to catch the typo in the resource limit, the missing namespace, or the wrong image tag. PR reviews are your last check before a change hits your cluster. A few practical tips:
```

- Enable branch protection rules requiring at least one approval before merge.
- Assign specific teams or individuals as CODEOWNERS for critical paths in your GitOps repo.

> - Pair human reviewers with CI checks (more on that in the next section).

Reviewer: The way this is written, it sounds like the next section will be about "pairing human reviewers", but it isn't — it's about leaning on CI checks. The "human" reference is to _this_ section; it's essentially saying, "combine what's in this section with what's in the next section", which is an awkward construct. 

```suggestion
- Catch as many issues as you can in CI to minimize the level of effort for reviews (more on this in the next section).
```

> - Beyond Git-level controls, use ArgoCD AppProjects to restrict which namespaces and clusters each team can target. A team should never be able to deploy to another team's namespace, even if they have push access to the repo. Flux achieves similar isolation through namespace-scoped Kustomization controllers and service account impersonation.

Reviewer: How does this work in practice? If I push to main, does Argo somehow see this, validate that `cnunciato` doesn't belong to some group, and reject the deployment? (And I guess if so, would a PR even work, since the committer would be the same with or without a code review? 🤔 )

{{% /notes %}}

## 8. Validate before you merge

Catching errors after they've been applied to your cluster is expensive. Catching them in CI before the merge is cheap. Shift left as hard as you can.

Your GitOps CI pipeline should validate everything it can before the change reaches your cluster: YAML syntax, Kubernetes schema validation, policy compliance, dry-run rendering.

- Tools like [`yamllint`](https://github.com/adrienverge/yamllint) and [`kubeconform`](https://github.com/yannh/kubeconform) catch syntax errors and schema violations before they become runtime failures. (Note: `kubeval`, which you'll see in older guides, is no longer maintained — its own README points to kubeconform as the replacement.)
- Run `helm template` or `kustomize build` in CI to verify that your templates render without errors.
- Use [OPA](https://www.openpolicyagent.org/)/[Conftest](https://www.conftest.dev/) or [Kyverno](https://kyverno.io/) to enforce organizational policies (no privileged containers, required labels, resource limits set) before merge.

> - If you're using the [rendered manifests pattern](/blog/gitops-best-practices-i-wish-i-had-known-before/#5-use-directories-not-branches-for-environments) from Section 5, your CI gets a bonus: PRs against the environment branches show the actual rendered YAML diff, not just a one-line Helm chart version bump hiding a thousand lines of manifest changes. One caveat — don't use SOPS or anything that renders plaintext secrets into the output. Stick with External Secrets Operator or similar reference-based approaches.

Reviewer: I'm actually confused by the inclusion of this bullet:

* How is this a bonus, as it relates to pre-merge validation? 
* Are PRs typically used with this pattern? Up to now, my impression is that with this pattern, commits get pushed to these branches directly and automatically, not after a human-approved PR. Is that right? Or did I get that wrong? And if they are, it's worth asking: Do we care? If the tip is about CI validation, does that change at all simply because you're using this pattern? (I wouldn't think so, but maybe it does?)
* The last bit about secrets exposure seems both important and unrelated to this tip. (This tip is about CI validation, whereas this is about taking care not expose secrets.) It feels like we should either uplevel the visibility of it (it's easy to miss tacked onto the end like this), remove it (because it's not strictly relevant to this tip), or give it its own tip. Thoughts?

Here's a minimal GitHub Actions workflow that validates your GitOps manifests on every PR:

```yaml
name: Validate GitOps Manifests
on:
  pull_request:
    paths: ["environments/**"]

jobs:
  validate:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - name: Validate YAML schemas
        run: kubeconform --strict environments/**/*.yaml
      - name: Build and verify Kustomize output
        run: |
          for env in environments/*/; do
            kustomize build "$env" > /dev/null
          done
      - name: Diff against live cluster
        run: |
          kustomize build environments/staging/ | \
            kubectl diff -f - || true
```

{{% notes type="info" %}}
Pro Tip: Add a `kubectl diff` step in your CI pipeline that shows exactly what would change in the cluster if the PR were merged. This gives reviewers a concrete view of the impact, not just the YAML diff.
{{% /notes %}}

## 9. Tag with commit SHAs, not "latest"

> Using `latest` as your image tag in a GitOps workflow is playing Russian roulette with your deployments. You have no idea what version is actually running, you can't reliably roll back, and your Git history becomes meaningless because the same manifest could deploy completely different code depending on when it was synced.

Reviewer: A bit unnecessarily gruseome. 😄 

```suggestion
Using `latest` as your image tag in a GitOps workflow is playing fast and loose with your deployments. You have no idea what version is actually running, you can't reliably roll back, and your Git history becomes meaningless because the same manifest could deploy completely different code depending on when it was synced.
```

- Tag your container images with the Git commit SHA that produced them. This creates a direct link between your source code, your build, and your deployment.
- Once a tag is pushed, it should never be overwritten. No re-tagging, no "oops let me push a fix to the same tag."
- Need to roll back? Just revert the commit that updated the image tag. The previous SHA still points to the exact same image it always did.

{{% notes type="info" %}}
Pro Tip: Set up an admission controller or CI policy that rejects any manifest using `latest` or any other mutable tag. Make it impossible to deploy without a pinned version.
{{% /notes %}}

> ## 10. Automate drift detection and reconciliation

Author: Done. Reframed the overall tip around "be precise about what you reconcile" as the core message.

Drift is when your cluster's actual state doesn't match the desired state in Git. It happens when someone runs a manual `kubectl` command, when an autoscaler changes a replica count, or when a CRD controller modifies a resource behind your back.

The whole point of GitOps is continuous reconciliation. Your operator should constantly compare the live state against Git and bring things back in line. If you've disabled auto-sync or set your reconciliation interval to "once a day," you're missing the point.

> - As mentioned earlier, a 1-minute reconciliation interval gives you high confidence that what's in Git is what's in your cluster.

Reviewer: Didn't we establish this as a bit of an anti-pattern?

- Configure your GitOps tool to automatically revert manual changes, but build your exclusion lists before you turn on auto-sync, not after. Controllers like Istio, cert-manager, and Crossplane legitimately modify resources, and auto-remediating those changes creates reconciliation loops that can destabilize a cluster. Start with a conservative exclusion list and tighten it over time.
- Even with auto-remediation, you want to know when drift happens. Set up alerts. It might indicate a process problem or a team member who needs coaching.

{{% notes type="info" %}}
Pro Tip: Be thoughtful about what you auto-remediate. HPAs, VPAs, cert-manager annotations, and external-dns records all legitimately modify resources. Use ignore rules to exclude fields that are intentionally dynamic. In ArgoCD, that looks like this:

```yaml
spec:
  ignoreDifferences:
    - group: apps
      kind: Deployment
      jsonPointers:
        - /spec/replicas
    - group: admissionregistration.k8s.io
      kind: MutatingWebhookConfiguration
      jqPathExpressions:
        - .webhooks[]?.clientConfig.caBundle
```

> Also, don't overlook resource ordering. ArgoCD sync waves and Flux's `dependsOn` let you enforce that CRDs are applied before custom resources and namespaces before workloads. Getting ordering wrong is one of the most common causes of failed syncs in production.

Author: Done. Added links to ArgoCD sync waves and Flux dependsOn documentation.

{{% /notes %}}

## 13. Secrets don't belong in Git (ever)

Git history is forever. Once a secret is committed, it lives in the repository's history even if you delete it in a subsequent commit. And if your GitOps repo is your single source of truth, you need a strategy for secrets that doesn't involve committing them in plaintext.

This is one of the biggest pain points in GitOps. The good news: there are several solid solutions:

- [External Secrets Operator](https://external-secrets.io/) syncs secrets from external providers (AWS Secrets Manager, HashiCorp Vault, [Pulumi ESC](/docs/esc/)) into Kubernetes secrets.
- [Sealed Secrets](https://sealed-secrets.netlify.app/) encrypts secrets client-side so only the controller in your cluster can decrypt them. Safe to commit the encrypted version to Git.
- [Pulumi ESC](/docs/esc/) manages secrets and configuration across environments with fine-grained access controls, and integrates with [Kubernetes via the External Secrets Operator](/docs/esc/integrations/kubernetes/external-secrets-operator/) or the [Secret Store CSI Driver](/docs/esc/integrations/kubernetes/secret-store-csi-driver/).

> - For a purely GitOps-native approach, [SOPS](https://github.com/getsops/sops) + age encrypts only the secret values inside your YAML files using `age` keys — the keys and file structure stay in plaintext, so diffs are still readable. Unlike Sealed Secrets, which introduces a `SealedSecret` CRD and a controller to decrypt it, SOPS works directly on your manifest files. No extra CRDs, no cluster-side operator. Just commit the encrypted files to Git. Widely used in the Flux ecosystem.

Reviewer: What is "age" here? Is this a tool, a value, something else?

{{% /notes %}}

## 14. Bridge your IaC and GitOps (don't choose one)

I hear this all the time: "We use Terraform for infrastructure and ArgoCD for deployments, and they live in completely separate worlds." Sound familiar? Most teams treat IaC and GitOps as independent workflows, but they're really two halves of the same story.

IaC handles "Day 0," creating the cloud resources your cluster depends on: VPCs, the Kubernetes cluster itself, IAM roles, databases, DNS zones. GitOps handles "Day 2," managing what runs inside the cluster: applications, addons, configurations. The gap between them is where things get messy. Your Helm charts need IAM role ARNs. Your GitOps-deployed addons need to know the cluster's OIDC provider endpoint. That metadata has to flow from the IaC layer to the GitOps layer somehow.

The [gitops-bridge](https://github.com/gitops-bridge-dev/gitops-bridge) pattern solves this cleanly: IaC provisions cloud resources and writes metadata (role ARNs, account IDs, endpoints) into resources that your GitOps tool can consume.

- Let each tool do what it's best at. Don't force Terraform's Helm provider to fight with ArgoCD over resource ownership.

> - [Pulumi](/docs/get-started/) lets you define your cloud infrastructure in real programming languages (TypeScript, Python, Go), which makes it natural to compute and pass metadata to your GitOps layer. The [Pulumi Kubernetes Operator](/docs/iac/guides/continuous-delivery/pulumi-kubernetes-operator/) even lets ArgoCD reconcile Pulumi stacks via GitOps.

Author: Done. Added stack outputs reference: "You can use stack outputs to expose values like role ARNs and cluster endpoints."

- Statsig went from "1-2 devs clicking around cloud consoles with SEVs left and right" to fully self-service by using [Pulumi to generate manifests and ArgoCD to deploy them](https://www.statsig.com/blog/scaling-infra-with-pulumi-argocd).

> - As you grow beyond a handful of clusters, think multi-cluster from the start. Patterns like ArgoCD ApplicationSets with cluster generators and Flux's Kustomization targeting become important for fleet management. The gitops-bridge pattern works well here because your IaC layer can register new clusters and write their metadata into Git, and ApplicationSets automatically pick them up.

Author: Done. Clarified to "write their metadata into Kubernetes resources" instead of "into Git."
